### PR TITLE
Converted Gen to be a sealed class; moved exhaustive and arb into thi…

### PR DESCRIPTION
**Pull request by sksamuel**
…s sealed space; simplified builders and gen methods

### CODE REVIEW

**Comment by sksamuel** [comment]
@dump247 interested in feedback. This is an implementation of arb and exhaustive as a sealed hierarchy.

**Comment by sksamuel** [comment]
The most important change is the `sealed class Gen` definition.

**Comment by ashishkujoy** [comment]
Looks good to me

File: kotest-property/src/commonMain/kotlin/io/kotest/property/Gen.kt [hunk 1]
```diff
@@ -61,11 +137,13 @@ fun <A> sampleOf(a: A, shrinker: Shrinker<A>) = Sample(a, shrinker.rtree(a))
...
-   override fun minIterations(): Int = this@merge.minIterations() + other.minIterations()
-   override fun generate(rs: RandomSource): Sequence<Sample<A>> {
-      return this@merge.generate(rs).zip(other.generate(rs)).flatMap {
-         sequenceOf(it.first, it.second)
-      }
-   }
-}
+fun <A, B : A> Gen<A>.merge(other: Gen<B>): Gen<A> = TODO()
+
+//object : Gen<A>() {
```
Comment (sschuberth) [inline kotest-property/src/commonMain/kotlin/io/kotest/property/Gen.kt:142]: Out-commented code could be removed.

**Comment by sksamuel** [comment]
Going to merge this in as it's blocking beta3 and there's been no further feedback. We can retroactiveily change / revert for beta4 if required.

**Pull request merged**